### PR TITLE
[stdlib] Slice: customize withContiguous[Mutable]StorageIfAvailable

### PR DESCRIPTION
**Pull request by lorentey**
We can easily make an `UnsafeBufferPointer` that slices another `UnsafeBufferPointer`, so let’s allow `Slice` to vend a slice of the base collection’s contiguous storage, if it provides access to one.

We need to do some index distance calculations to implement this, but those will be constant-time in the usual case where the base collection is a RAC.

https://bugs.swift.org/browse/SR-11957
rdar://58090587

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

**Review by stephentyrone — APPROVED** [review]

File: validation-test/stdlib/UnsafeBufferPointer.swift.gyb [hunk 1]
```diff
@@ -361,6 +361,32 @@ UnsafeMutableBufferPointerTestSuite.test("withContiguous(Mutable)StorageIfAvaila
   expectEqual(result1, result2)
 }
 
+UnsafeMutableBufferPointerTestSuite.test("Slice.withContiguousStorageIfAvailable") {
+  guard #available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *) else {
+    expectTrue(false)
```
Comment (lorentey) [inline validation-test/stdlib/UnsafeBufferPointer.swift.gyb]: Oops, this slipped in from my local testing. I'll force push to get rid of it when the first test run completes,

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Review by natecook1000 — APPROVED** [review]
Looks great!

**Pull request merged**